src/client/chat.js

```diff
@@ -0,0 +1,538 @@
+            acknowledgements[idx] = { signature: entry.signature, pending: false }
+          }
+        }
+
+        const timestamp = BigInt(Date.now())
+        const salt = 0n
+        const bitset = Buffer.allocUnsafe(3)
+        bitset[0] = ackSet & 0xFF
+        bitset[1] = (ackSet >> 8) & 0xFF
+        bitset[2] = (ackSet >> 16) & 0xFF
+
+        client.write('chat_message', {
+          message,
+          timestamp,
+          salt,
+          signature: client.signMessage(message, timestamp, salt, toSign.map(ack => ack.signature)),
+          offset,
+          acknowledged: bitset
+        })
+      } else {
+        const timestamp = BigInt(Date.now())
+        client.write('chat_message', {
+          message,
+          timestamp,
+          salt: 0n,
+          signature: client.signMessage(message, timestamp)
+        })
+      }
+    } else client.write('chat', { message })
+  }
+
+  function fetchPreview (message) {
+    return new Promise((resolve, reject) => {
+      if (client.pendingPreview) {
+        if (options.queueMessages) queueMessagePreview(message, resolve, reject)
+        else reject(Error('Message could not be sent. Preview already in progress'))
+      } else {
+        requestPreview(message, resolve, reject)
+      }
+    })
+  }
+
+  function requestPreview (message, resolve, reject) {
+    client.previewId = client.previewId || 0
+    const id = client.previewId++
+    client.pendingPreview = { id, resolve, reject }
+
+    if (!client.lastPreviewRequest) client.lastPreviewRequest = 0
+    const diff = Date.now() - client.lastPreviewRequest
+
+    function sendPacket () {
+      client.write('chat_preview', {
+        query: id,
+        message
+      })
+
+      if (options.timeoutPreview) {
+        client.previewTimeout = setTimeout(() => {
+          throw Error('Server did not respond with preview in time') // Probably should not throw
+        }, options.timeoutPreview)
+      }
+    }
+
+    if (diff < 100) {
+      setTimeout(sendPacket, diff + 1)
+    } else sendPacket()
+  }
+
+  function queueMessagePreview (message, resolve, reject) {
+    if (!client.previewQueue) client.previewQueue = []
+    client.previewQueue.push({ message, resolve, reject })
+  }
+
+  // Packet handling
+
+  client.on('chat_preview', handleChatPreview)
+
+  client.on('player_info', handlePlayerInfo)
+
+  client.on('player_remove', handlePlayerRemove)
+
+  client.on('player_chat', handlePlayerChat)
+
+  client.on('system_chat', handleSystemChat)
+
+  client.on('hide_message', handleHideChat)
+
+  client.on('message_header', handleChatHeader)
+
+  client.on('should_display_chat_preview', packet => {
+    client.useChatPreview = packet.should_display_chat_preview
+  })
+
+  client.on('server_data', packet => {
+    client.useChatPreview = packet.previewsChat
+    client.signedChat = packet.enforcesSecureChat
+  })
+
+  function handleChatPreview (packet) {
+    if (packet.queryId === client.pendingPreview.id) {
```

Comment:
the previews are not necessarily synchronously ordered, for example you might send two previews # 1 and # 2 and the server can respond in the order of # 2 and # 1

Reply:
While this is true, the vanilla client makes it virtually impossible to send 2 preview requests at once (essentially by treating responses the same way I do), and therefore I have not added that possibility either

Reply:
You're getting messages here from the server and the client implements the ticketing system on purpose to avoid asynchronous problems. Since the protocol is designed asynchronously it's not inconceivable that flooding messages to a server and then getting previews back would come out of order, depending on how the server is processing the message.

Reply:
My point is the vanilla client does not support that either. The client only awaits one preview at once. If they come out of order the client will ignore them

Reply:
Can you link to the source code for that?

Reply:
https://github.com/extremeheat/extracted_minecraft_data/blob/client1.19.1/client/net/minecraft/client/gui/chat/ChatPreviewRequests.java

https://github.com/extremeheat/extracted_minecraft_data/blob/client1.19.1/client/net/minecraft/client/gui/chat/ClientChatPreview.java

ChatPreviewRequests only tracks one message at a time. ClientChatPreview only tracks one ChatPreviewRequests. It does attempt to send out more previews every 100ms but when the response comes back it only compares to the latest. If the preview response is not the latest it just returns null